scikeras/wrappers.py

```diff
@@ -888,84 +876,64 @@ def preprocess_y(
         meta : Dict[str, Any]
             Meta parameters, as determined from input.
         """
-        y, meta = super().preprocess_y(y, reset=reset)
+        y, meta = super().preprocess_y(y)
 
         loss = self.loss
 
-        if self.target_type_ == "binary":
-            # y = array([1, 0, 1, 0])
-            # single task, single label, binary classification
-            # encode
-            if reset:
-                encoder = make_pipeline(
-                    Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
-                ).fit(y)
-                meta.update(
-                    {
-                        "classes_": encoder[1].categories_[0],
-                        "n_classes_": encoder[1].categories_[0].size,
-                        "target_encoder_": encoder,
-                        "model_n_outputs_": 1,
-                        "n_outputs_": 1,
-                    }
-                )
-            else:
-                encoder = self.target_encoder_
-            y = encoder.transform(y)
-        elif self.target_type_ == "multiclass":
-            # y = array([1, 5, 2])
-            # encode
-            if reset:
-                if is_categorical_crossentropy(loss):
-                    # one-hot encode
-                    encoder = make_pipeline(
-                        Ensure2DTransformer(),
-                        OneHotEncoder(sparse=False, dtype=np.float32),
-                    )
-                else:
-                    encoder = make_pipeline(
-                        Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
-                    )
-                y = encoder.fit_transform(y)
-                meta.update(
-                    {
-                        "classes_": encoder[1].categories_[0],
-                        "n_classes_": encoder[1].categories_[0].size,
-                        "target_encoder_": encoder,
-                        "model_n_outputs_": 1,
-                        "n_outputs_": 1,
-                    }
-                )
-            else:
-                encoder = self.target_encoder_
-                y = encoder.transform(y)
+        encoders = {
+            "binary": make_pipeline(
+                Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
+            ),
+            "multiclass": make_pipeline(
+                Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
+            ),
+            "multiclass-one-hot": FunctionTransformer(),
+        }
```

Comment:
We instantiate new transformers and call `make_pipeline` every single time even if a couple lines down we ignore all of that work with `if not hasattr(self, "target_encoder_")`. I would move all of this within `if not hasattr(self, "target_encoder_").